keps/sig-network/20190423-service-lb-finalizer.md

```diff
@@ -0,0 +1,156 @@
+seems to be common enough and is worth to have a better mechanism for ensuring
+the cleanup of load balancing resources.
+
+### Goals
+
+Ensure the Service resource is not fully deleted until the correlating load
+balancing resources are also deleted.
+
+## Proposal
+
+We are going to define a finalizer for service LoadBalancers with name
+`kubernetes.io/service-load-balancer`. This finalizer will be attached to any
+service that has `type=LoadBalancer`. Upon the deletion of such servicce, the
+actual deletion of the resource will be blocked until this finalizer is removed.
+This finalizer will not be removed until cleanup of the correlating load
+balancing resources are considered finished by service controller.
+
+Note that the removal of this finalizer might also happen when service type
+changes from `LoadBalancer` to another. This however doesn't change the
+implication that the resources cleanup must be fulfilled before fully deleting
+the service.
+
+The lifecyle of a `LoadBalancer` type serivce with finalizer would look like:
+- Creation
+  1. User creates a service.
+  2. Service controller observes the creation and attaches finalizer to the service.
+  3. Provision of load balancing resources.
+- Deletion
+  1. User issues a deletion for the service.
+  2. Service resoure deletion is blocked due to the finalizer.
+  3. Service controller observed the deletion timestamp is added.
+  4. Cleanup of load balancing resources.
+  5. Service controller removes finalizer from the service.
+  6. Service resource deleted.
+- Update to another type
+  1. User update service from `type=LoadBalancer` to another.
+  2. Service controller observed the update.
+  3. Cleanup of load balancing resources.
+  4. Service controller removes finalizer from the service.
+
+The expected cluster upgrade/downgrade path for service with finalizer would be:
+- Upgrade from pre-finalizer version
+  - All existing `LoadBalancer` services will be attached a finalzer upon startup
+  of the new version of service controller.
+  - The newly created `LoadBalancer` services will have finalizer attached upon
+  creation.
+- Downgrade from with-finailzer version
+  - All existing `LoadBalancer` service will have the attached finalizer removed
+  upon the cleanup of load balancing resources.
+  - The newly created `LoadBalancer` services will not have finailzer attached.
+
+To ensures that downgrades can happen safely, the first release will include the
+"remove finalizer" logic with the "add finalizer" logic behind a gate. Then in a
+later release we will remove the feature gate and enable both the "remove" and
+"add" logic by default.
+
+As such, we are proposing Alpha/Beta/GA phase for this enhancement as below:
+- Alpha: The finailizer cleanup logic would be on by default while the finalizer
+addition logic will be put behind a feature gate that is off by default.
+- Beta: Both the cleanup and addition logic will be default on, but allowing
```

Comment:
Should this be: 
```
- Beta: Finalizer clean up will always be on. Finalizer addition will be on by default but can be disabled via a feature gate.
```
instead?

Reply:
I think he's proposing this state for Alpha, and I think it's probably OK.  

We could require:
* pre-alpha: cleanup and addition gated, default off
* alpha: cleanup gated, default on; addition gated default off
* beta: cleanup and addition gated, default on
* GA: default on

That seems tedious.  The risk of dropping pre-alpha is that people come to depend on the cleanup but we might remove it OR that cleanup causes real problems.  Do we think that's signficant?  I think it's a calculated risk that shaves a whole release off such a feature...

Reply:
I guess what I meant was exactly what @andrewsykim suggested :)

We were trying to avoid the pre-alpha phase --- having the feature available but completely off by default. I can't easily think of how the cleanup can cause significant problem, as that is trying to remove a specific finalizer that presumably doesn't exist yet.

> people come to depend on the cleanup

Not sure if there is other use case, we wanted to enable the cleanup but not the addition just to accommodate the downgrade problem.

Reply:
> We were trying to avoid the pre-alpha phase --- having the feature available but completely off by default. I can't easily think of how the cleanup can cause significant problem, as that is trying to remove a specific finalizer that presumably doesn't exist yet. 

Agreed, I think something like this is more ideal (and I think the KEP is supposed to reflect this?):
* pre-alpha / alpha: clean up always on, addition gated with default off 
* beta: clean up always on, addition gated default on
* GA: clean up & addition always on

> The risk of dropping pre-alpha is that people come to depend on the cleanup but we might remove it OR that cleanup causes real problems.

The biggest risk I think is in trying to change the clean-up logic. If we change it in any breaking way during alpha, we have to wait another release cycle to account for rollbacks. Two releases _if_ we want to account for those users who like to be dangerous and do n+2 version upgrades.

Reply:
I thought that sequence was what @MrHohn was proposing?

As to waiting another release - that is easy...don't screw up :)

Reply:
> I thought that sequence was what @MrHohn was proposing?

Sorry, I was a bit confused earlier. I think we're on the same page now :)